compiler/rustc_lint/src/types.rs

```diff
@@ -1144,10 +1144,6 @@ pub(crate) fn repr_nullable_ptr<'tcx>(
             [var_one, var_two] => match (&var_one.fields.raw[..], &var_two.fields.raw[..]) {
                 ([], [field]) | ([field], []) => field.ty(tcx, args),
                 ([field1], [field2]) => {
-                    if !tcx.features().result_ffi_guarantees {
-                        return None;
-                    }
```

Comment:
Note that this makes the lint accept all "Result-like" enums, not just `Result` specifically. This goes beyond the RFC and beyond what we have documented. It matches what the lint does on `Option`, though. I don't know why the lint does this, and could not find a comment explaining this choice.

Reply:
The logic here now does not quite match what has been FCP'd, so the lint will fire on some cases where it doesn't have to fire -- specifically if the "other" variant (the one that does not contain the non-null type) has more than one field but all fields are stably 1-ZST.

This PR can land without fixing that though, it just means the lint can be improved, but that's already well-established. ;)

(I also find `is_niche_optimization_candidate` a very confusing name for "test if this is a 1-ZST and will remain a 1-ZST under semver updates". This type doesn't even have any niches... at least the function has a doc comment clearly describing what it does.)

Reply:
I lifted this comment into #72774 as part of that refactoring.